enhancements/update/upgrade-without-registry.md

```diff
@@ -0,0 +1,893 @@
+available space for these images, as otherwise the performance of the node may
+degrade and it may stop functioning correctly if it runs out of space. The space
+should be enough to accommodate tho releases (current running + candidate for
+install) as well as workload images and buffer.
+
+#### Don't try to contact the image registry server explicitly
+
+Some OpenShift components explicitly try to contact the registry server without
+a fallback alternative. These need to be changed so that they don't do it or so
+that they have a fallback mechanism when the registry server isn't available.
+
+For example, in OpenShift 4.1.13 the machine config operator runs the equivalent
+of `skopeo inspect` in order to decide what kind of upgrade is in progress. That
+fails if there is no registry server, even if the image has already been pulled.
+That needs to be changed so that contacting the registry server is not required.
+A possible way to do that is to use the equivalent of `crictl inspect` instead.
+
+#### MCO should support the configuration changes required for the upgrade
+
+In order to copy the images required for the upgrade to the nodes of the cluster
+we will create an additional image store in the `/var/lib/upgrade` directory of
+each node of the cluster, and we will pin all those images. This requires
+changes in the `/etc/containers/storage.conf` file, something like this:
+
+```toml
+additionalimagestores = [
+  "/var/lib/upgrade/4.13.7-x86_64"
+]
+```
+
+That `/etc/containers/storage.conf` file is tracked by the machine config
+operator, and changing it will trigger a reboot that will interfere with the
+upgrade process. We will need to change the machine config operator so that it
+is aware of these changes and doesn't reboot the node. Ideally this should be
+added to the `ContainerRuntimeConfig` object, for example:
+
+```yaml
+apiVersion: machineconfiguration.openshift.io/v1
+kind: ContainerRuntimeConfig
+metadata:
+  ...
+spec:
+  containerRuntimeConfig:
+    additionalImageStores:
+    - /var/lib/upgrade/4.13.7-x86_64
+```
+
+When the new `additionalImageStores` field is added or changed the machine
+config operator will need to regenerate the `/etc/containers/storage.conf` file,
+including the corresponding `additionalimagestores` field, and then reload it
+using the equivalent of `systemctl reload crio.service`.
+
+The changes to pin the images will be done in a
+`/etc/crio/crio.conf.d/pin-upgrade.conf` file, something like this:
+
+```toml
+pinned_images=[
+  "quay.io/openshift-release-dev/ocp-release@sha256:...",
+  "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+  "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+  ...
+]
+```
+
+These files aren't tracked by the machine config operator, but the CRI-O service
+needs to be reloaded when they are created or changed. To support that a new
+filed will be added to the `ContainerRuntimeConfig` object, for example:
```

Comment:
What creates this new resource in the cluster for each bundle?

Reply:
The cluster version operator.